File: Ruby/linked_list.rb
class Node
  attr_accessor :value, :next_node

  def initialize(value, next_node = nil)
    @value = value
    @next_node = next_node
  end
end

class LinkedList
  attr_reader :head, :tail

  def initialize
    @head = @tail = nil
  end

  def add(number)
    new_node = Node.new(number)
    if @head.nil?
      @head = @tail = new_node
    else
      @tail.next_node = new_node
      @tail = new_node
    end
  end

  def get(index)
    current_node = @head
    index.times { current_node = current_node.next_node }
    current_node.value
  end

  def add_at(index, item)
    new_node = Node.new(item)
    if @head.nil?
      @head = @tail = new_node
    elsif index.zero?
      new_node.next_node = @head
      @head = new_node
    elsif index.positive?
      current_node = @head
      (index - 1).times { current_node = current_node.next_node }
      new_node.next_node = current_node.next_node
      @tail = new_node if current_node.next_node.nil?
      current_node.next_node = new_node
    end
  end

  def remove(index)
    current_node = @head
    return index if @head == @tail

    if index.zero?
      @head = @head.next_node
    elsif index.positive?
      (index - 2).times { current_node = current_node.next_node }
      current_node.next_node = current_node.next_node.next_node
    end
  end

  def show_list
    current_node = @head
    until current_node.nil?
      puts current_node.value
      current_node = current_node.next_node
    end
  end
end

File: Typescript/DataStructures/linkedList.ts
class Node_LinkedList {
  value: number | string;
  next: Node_LinkedList | null;
  constructor(data: number | string) {
    this.value = data;
    this.next = null;
  }
}

class LinkedList {
  head: Node_LinkedList | null;
  constructor() {
    this.head = null;
  }

  append(value: number | string) {
    const newNode: Node_LinkedList = new Node_LinkedList(value);
    if (!this.head) {
      this.head = newNode;
    } else {
      let currentNode: Node_LinkedList = this.head;
      while (currentNode.next) {
        currentNode = currentNode.next;
      }
      currentNode.next = newNode;
    }
  }

  prepend(value: number | string) {
    const newNode: Node_LinkedList = new Node_LinkedList(value);
    newNode.next = this.head;
    this.head = newNode;
  }

  remove(value: number | string) {
    if (!this.head) return;
    if (this.head.value === value) {
      this.head = this.head.next;
      return;
    }
    let currentNode: Node_LinkedList = this.head;
    while (currentNode.next && currentNode.next.value !== value) {
      currentNode = currentNode.next;
    }
    if (currentNode.next) currentNode.next = currentNode.next.next;
  }

  print() {
    let currentNode: Node_LinkedList = this.head;
    const values = [];
    while (currentNode) {
      values.push(currentNode.value);
      currentNode = currentNode.next;
    }
    console.log({
      "Current List: ": values.join(` => `),
    });
  }
}

const newList = new LinkedList();
newList.append(1);
newList.append(10);
newList.prepend(0);
newList.prepend(20);
newList.append(30);
newList.print();
newList.remove(10);
newList.print();

console.log(newList);
